components/core/src/os/net/mod.rs

```diff
@@ -22,3 +22,73 @@ mod imp;
 mod imp;
 
 pub use self::imp::*;
+
+extern crate dns_lookup;
+
+use dns_lookup::{getaddrinfo,
+                 AddrInfoHints};
+use std::io;
+
+// lookup_fqdn returns a vector of fqdn that resolves the provided hostname.
+//
+// Since the underlying crate is platform agnostic, this function is as well,
+// we only implement a single variable `flags` that has the right hint for
+// the running operating system.
+pub fn lookup_fqdn(hostname: &str) -> io::Result<String> {
+    #[cfg(not(windows))]
+    let flags = libc::AI_CANONNAME;
+
+    #[cfg(windows)]
+    let flags = winapi::shared::ws2def::AI_CANONNAME;
+
+    let hints = AddrInfoHints { flags,
+                                ..AddrInfoHints::default() };
+    let addrinfos =
+        getaddrinfo(Some(hostname), None, Some(hints))?.collect::<std::io::Result<Vec<_>>>()?;
+
+    // If 'hints.ai_flags' includes the AI_CANONNAME flag, then the ai_canonname
+    // field of the first of the addrinfo structures in the returned list is set
+    // to point to the official name of the host.
+    if !addrinfos.is_empty() {
+        let addrinfo = addrinfos[0].clone();
+        return Ok(addrinfo.canonname.unwrap_or_else(|| hostname.to_string()));
+    }
+
+    Ok(hostname.to_string())
+}
```

Comment:
If you make `addrinfos` above `mut`, then you can replace this bit with the following:

```rust
    let name = if let Some(addrinfo) = addrinfos.pop() {
        addrinfo.canonname.unwrap_or_else(|| hostname.to_string())
    } else {
        hostname.to_string()
    };
    Ok(name)
```
That simplifies things by removing a clone, and also removes a `return` call.

I think there are maybe some additional simplifications we can make (we ought to be able to just use `unwrap` or `expect`), but I'd want to do some additional testing first. This is good for now, though.

Reply:
There's no need to do a `collect` on the return of `getaddrinfo` at all, I don't think. I would suggest:
```rust
    if let Some(Ok(first_addr_info)) = getaddrinfo(Some(hostname), None, Some(hints))?.next() {
        Ok(first_addr_info.canonname
                       .expect("Some(canonname) if requested"))
    } else {
        Ok(hostname.to_string())
    }
```
or
```rust
    if let Some(first_result) = getaddrinfo(Some(hostname), None, Some(hints))?.next() {
        Ok(first_result?.canonname
                        .expect("Some(canonname) if requested"))
    } else {
        Ok(hostname.to_string())
    }
```
Depending on whether we want an `Err` variant of [`Result<AddrInfo>`](https://docs.rs/dns-lookup/1.0.0/dns_lookup/struct.AddrInfoIter.html) to return a `Result::Err` or the `hostname`.

Reply:
Updated!

Reply:
Reading through the [source code](https://docs.rs/dns-lookup/1.0.0/src/dns_lookup/addrinfo.rs.html#96) it looks like errors of the kind returned by `first_result?` would result from null pointers or non-v4 or v6 socket families (things that should be exceedingly rare and pathological).

I think we could probably log those errors and return `Ok(hostname.to_string())`

At that point, the only error this function would be returning would be the case where the lookup call itself fails. Over in the manager, we currently up falling back to the hostname from the `Sys` struct. That hostname comes from our existing homegrown `hostname` implementation, which I now think that we should replace with the one from `dns_lookup` for consistency. That can be a followup PR, though.

Reply:
I'll try to update it to log a debug message and return the hostname as you mentioned.